searcher_test.go

```diff
@@ -0,0 +1,162 @@
+			Cert:      cert("etcd"),
+			Secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterID":        "eggs2",
+						"clusterComponent": "etcd",
+					},
+				},
+				Data: map[string][]byte{
+					"ca":  []byte("test-CA"),
+					"crt": []byte("test-crt"),
+					"key": []byte("test-key"),
+				},
+			},
+			ExpectedTLS: TLS{
+				CA:  []byte("test-CA"),
+				Crt: []byte("test-crt"),
+				Key: []byte("test-key"),
+			},
+			ExpectedErrMatch: isNil,
+		},
+		// 1: cluster ID doesn't match.
+		{
+			ClusterID: "eggs5",
+			Cert:      cert("etcd"),
+			Secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterID":        "eggs2",
+						"clusterComponent": "etcd",
+					},
+				},
+				Data: map[string][]byte{
+					"ca":  []byte("test-CA"),
+					"crt": []byte("test-crt"),
+					"key": []byte("test-key"),
+				},
+			},
+			ExpectedTLS:      TLS{},
+			ExpectedErrMatch: IsInvalidSecret,
+		},
+		// 2: cert doesn't match.
+		{
+			ClusterID: "eggs2",
+			Cert:      cert("calico"),
+			Secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterID":        "eggs2",
+						"clusterComponent": "etcd",
+					},
+				},
+				Data: map[string][]byte{
+					"ca":  []byte("test-CA"),
+					"crt": []byte("test-crt"),
+					"key": []byte("test-key"),
+				},
+			},
+			ExpectedTLS:      TLS{},
+			ExpectedErrMatch: IsInvalidSecret,
+		},
+		// 3: ca field missing.
+		{
+			ClusterID: "eggs2",
+			Cert:      cert("etcd"),
+			Secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterID":        "eggs2",
+						"clusterComponent": "etcd",
+					},
+				},
+				Data: map[string][]byte{
+					"crt": []byte("test-crt"),
+					"key": []byte("test-key"),
+				},
+			},
+			ExpectedTLS:      TLS{},
+			ExpectedErrMatch: IsInvalidSecret,
+		},
+		// 4: crt field missing.
+		{
+			ClusterID: "eggs2",
+			Cert:      cert("etcd"),
+			Secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterID":        "eggs2",
+						"clusterComponent": "etcd",
+					},
+				},
+				Data: map[string][]byte{
+					"ca":  []byte("test-CA"),
+					"key": []byte("test-key"),
+				},
+			},
+			ExpectedTLS:      TLS{},
+			ExpectedErrMatch: IsInvalidSecret,
+		},
+		// 5: key field missing.
+		{
+			ClusterID: "eggs2",
+			Cert:      cert("etcd"),
+			Secret: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterID":        "eggs2",
+						"clusterComponent": "etcd",
+					},
+				},
+				Data: map[string][]byte{
+					"ca":  []byte("test-CA"),
+					"crt": []byte("test-crt"),
+				},
+			},
+			ExpectedTLS:      TLS{},
+			ExpectedErrMatch: IsInvalidSecret,
+		},
+	}
+
+	for i, tc := range testCases {
+		var tls TLS
+		err := fillTLSFromSecret(&tls, tc.Secret, tc.ClusterID, tc.Cert)
```

Comment:
I personally do not like working with pointers that way. It is more straight forward to call a function to get some value back.

Reply:
I also favour returning values over passing pointers and mutating them. I just couldn't make better design. This is a private function - we can change it any time. Please feel free to provide a PR.